higgstautau/lephad/filters.py

```diff
@@ -0,0 +1,403 @@
+        self.min_leps = min_leps
+        self.el_thresh = el_thresh
+        self.mu_thresh = mu_thresh
+        super(LepPT, self).__init__(**kwargs)
+    def passes(self, event):
+        event.electrons.select(lambda electron: electron.cl_E/cosh(electron.tracketa) > self.el_thresh)
+        event.muons.select(lambda muon: muon.pt > self.mu_thresh)
+        return len(event.electrons) + len(event.muons) >= self.min_leps
+
+class LepEta(EventFilter):
+    def __init__(self, min_leps, **kwargs):
+        self.min_leps = min_leps
+        super(LepEta, self).__init__(**kwargs)
+    def passes(self, event):                                                                                         
+        # both calo and leading track eta within 2.47                                                                
+        event.electrons.select(lambda electron: 
+            abs(electron.cl_eta) < 1.37 or
+            1.52 < abs(electron.cl_eta) < 2.47)
+        event.muons.select(lambda muon: abs(muon.eta) < 2.5)
+        return len(event.electrons) + len(event.muons) >= self.min_leps
+
+class LepHad(EventFilter):
+    def __init__(self, **kwargs):
+        super(LepHad, self).__init__(**kwargs)
+    def passes(self, event):
+        #make sure it's only one had and one lep
+        #sort in decending order by pT
+        event.electrons.sort(key=lambda electron: electron.cl_E/cosh(electron.tracketa), reverse=True)
+        event.muons.sort(key=lambda muon: muon.pt, reverse=True)
+        event.taus.sort(key=lambda tau: tau.pt, reverse=True)
+        if len(event.taus) != 1: 
+            return False
+        if len(event.electrons) == 1:
+            return True
+        elif len(event.muons) == 1:
+            return True
+        else: 
+            return False
+
+#from here on we get specific muon and electron filters.
+
+class MuonID(EventFilter):
+    def __init__(self, min_leps, **kwargs):
+        self.min_leps = min_leps
+        super(MuonID, self).__init__(**kwargs)
+    def passes(self, event):
+        #misc muon cuts
+        event.muons.select(lambda muon: muon.tight)
+        return len(event.electrons) + len(event.muons) >= self.min_leps
+
+class MuonHits(EventFilter):
+    def __init__(self, min_leps, **kwargs):
+        self.min_leps = min_leps
+        super(MuonHits, self).__init__(**kwargs)
+    def passes(self, event):
+        # Bottom layer requirements 
+        event.muons.select(lambda muon:
+            muon.expectBLayerHit == 0 or muon.nBLHits > 0)
+        # Pixel hits
+        event.muons.select(lambda muon:
+            (muon.nPixHits + muon.nPixelDeadSensors) > 0)
+        # SCT hits
+        event.muons.select(lambda muon:
+            (muon.nSCTHits + muon.nSCTDeadSensors) > 0)
+        # Holes
+        event.muons.select(lambda muon:
+            (muon.nPixHoles + muon.nSCTHoles) < 3)
+        # TRT
+        event.muons.select(lambda muon:
```

Comment:
Wow, I would define a small function def trt_hits(muon) that returns True or False and use it:
event.muon.select(lambda muon: trt_hits(muon)). 
The function could be defined within the filter line 91